probe-rs-tools/src/bin/probe-rs/cmd/dap_server/server/logger.rs

```diff
@@ -0,0 +1,224 @@
+
+    fn make_writer(&self) -> Self::Writer {
+        // The API doesn't allow graceful exit, but we do not expect locking of the Mutex to fail.
+        // If it does, we want to know about it.
+        #[allow(clippy::expect_used)]
+        self.buffer_file_handle()
+            .expect("Failed to get a lock on the file used to buffer tracing output.")
+    }
+}
+
+impl DebugLogger {
+    /// Create a new DebugTraceFile instance
+    pub(crate) fn new(log_file: Option<&Path>) -> Result<Self, DebuggerError> {
+        let mut debug_logger = Self {
+            buffer_file: Mutex::new(tempfile()?),
+            seek_pointer: SeekFrom::Start(0),
+            log_default_guard: None,
+        };
+        debug_logger.log_default_guard = Some(debug_logger.setup_logging(log_file)?);
+
+        Ok(debug_logger)
+    }
+
+    /// Try to clone the buffer file handle with exclusive access, so that it can be used by either
+    /// the DebugAdapter to send messages to the DAP client, or the `tracing` module to record log data.
+    pub(crate) fn buffer_file_handle(&self) -> Result<File, DebuggerError> {
+        self.buffer_file
+            .lock()
+            .map_err(|error| {
+                DebuggerError::Other(anyhow!(
+                    "Failed to get a lock on the buffer file. {error:?}"
+                ))
+            })?
+            .try_clone()
+            .map_err(|_| {
+                DebuggerError::Other(anyhow!(
+                    "Failed to get a new file instance to the buffer file."
+                ))
+            })
+    }
+
+    /// Flush the buffer to the DAP client's Debug Console
+    pub(crate) fn flush_to_dap(
+        &mut self,
+        debug_adapter: &mut DebugAdapter<impl ProtocolAdapter>,
+    ) -> Result<(), DebuggerError> {
+        let read_from_log = self.buffer_file_handle()?;
+        let mut tracing_log_handle = BufReader::new(read_from_log.try_clone()?);
+        std::io::Seek::seek(&mut tracing_log_handle, self.seek_pointer)?;
+        let mut buffer_lines = tracing_log_handle.lines();
+        while let Some(Ok(next_line)) = buffer_lines.next() {
+            debug_adapter.log_to_console(next_line);
+        }
+        // Update the seek_pointer to the end of the file, so that we don't re-read the same lines.
+        let mut truncate_file = read_from_log.try_clone()?;
+        let later_read_pos = std::io::Seek::seek(&mut truncate_file, std::io::SeekFrom::End(0))?;
+        self.seek_pointer = std::io::SeekFrom::Start(later_read_pos);
+        Ok(())
+    }
+
+    /// Flush the buffer to the stderr
+    pub(crate) fn flush(&mut self) -> Result<(), DebuggerError> {
+        let read_from_log = self.buffer_file_handle()?;
+        let mut tracing_log_handle = BufReader::new(read_from_log.try_clone()?);
+        std::io::Seek::seek(&mut tracing_log_handle, self.seek_pointer)?;
+        let mut buffer_lines = tracing_log_handle.lines();
+        while let Some(Ok(next_line)) = buffer_lines.next() {
```

Comment:
I think I understand what this wants to be, but I also think this is worth factoring out into its own function. The difference between `flush` and `flush_to_dap` is (or... seems to be) what happens with the log lines, and that single line could be passed in as a closure.

Reply:
I tried that, but it means that I have to pass an `Option<DebugAdapter<P>>` to the combined `flush` function. That isn't a problem in itself, but the compiler wants you to add a type when you pass a `None` value, and for the life of me I couldn't make the syntax work.  If you can make it work (you probably can), I'd appreciate the solution.

Reply:
`None::<type>` :)

Reply:
That being said I'm not sure we understand each other here, so this can go in as-is, and we can tackle this part later.

Reply:
It requires all the generics in the type, and I can't remember what the error was, but it doesn't like `None::<DapAdapter<P>>` ... or any of the variants I could come up with :)

Reply:
As I've said, I'm not sure at all what you're planning or why it needs a None. If you pass in a closure as I've suggested, the closure captures `debug_adapter` for you if you need to use it.